### PR TITLE
ChannelFileSendWithComplex

### PR DESCRIPTION
**Pull request by legolord208**
What's a better way of doing this? Reflect? Generator? I'm open to suggestions

### CODE REVIEW

**Comment by iopred** [comment]
Just put an io.Reader as an arg in SendMessageParams? We don't need another complex method.

**Comment by legolord208** [comment]
I would agree, but it uses a completely different `request` method, and content type.

**Comment by iopred** [comment]
Then message send complex needs to get a bit more complex :)

**Comment by legolord208** [comment]
@iopred You mean detecting if file is set, and if so use a completely different method? o_o

Though, how would the name be set? Have a file struct?

**Comment by iopred** [comment]
We're likely to have a SetFile method on SendParams like we did on edit. So
I think it's fine to have Filename and FileReader on the params

**Comment by legolord208** [comment]
I'll see what I can do tomorrow

**Comment by legolord208** [comment]
@iopred :bell:

File: restapi.go [hunk 1]
```diff
@@ -1311,7 +1311,49 @@ func (s *Session) ChannelMessageSendComplex(channelID string, data *MessageSend)
...
+		bodywriter := multipart.NewWriter(body)
+
+		// What's a better way of doing this? Reflect? Generator? I'm open to suggestions
+
+		if err = bodywriter.WriteField("content", data.Content); err != nil {
+			return nil, err
+		}
+		var embed []byte
+		embed, err = json.Marshal(data.Embed)
+		if err != nil {
```
Comment (iopred) [inline restapi.go]: You're only writing the embed on failure, this embed write shouldn't be in an if scope.

Comment (legolord208) [inline restapi.go]: Oh oops! Nice catch!

File: restapi.go [hunk 2]
```diff
@@ -1311,7 +1311,52 @@ func (s *Session) ChannelMessageSendComplex(channelID string, data *MessageSend)
...
+			return
+		}
+
+		err = bodywriter.Close()
+		if err != nil {
+			return
+		}
+
+		endpoint := EndpointChannelMessages(channelID)
```
Comment (iopred) [inline restapi.go]: Sorry I'm being really pedantic here, could you move this up outside the if, and use it in the else below as well, I'm being all weird and finicky :)

Comment (legolord208) [inline restapi.go]: No problem! Fixed!

File: restapi.go [hunk 3]
```diff
@@ -1311,7 +1311,53 @@ func (s *Session) ChannelMessageSendComplex(channelID string, data *MessageSend)
...
+	endpoint := EndpointChannelMessages(channelID)
+
+	var response []byte
+	if data.File != nil {
+		body := &bytes.Buffer{}
+		bodywriter := multipart.NewWriter(body)
+
+		// What's a better way of doing this? Reflect? Generator? I'm open to suggestions
+
+		if err = bodywriter.WriteField("content", data.Content); err != nil {
```
Comment (iopred) [inline restapi.go]: You should only do this if content is != ""

File: restapi.go [hunk 4]
```diff
@@ -1311,7 +1311,53 @@ func (s *Session) ChannelMessageSendComplex(channelID string, data *MessageSend)
...
+			return
+		}
+
+		var embed []byte
+		embed, err = json.Marshal(data.Embed)
+		if err != nil {
+			return
+		}
+		err = bodywriter.WriteField("embed", string(embed))
```
Comment (iopred) [inline restapi.go]: Only if embed != nil

File: restapi.go [hunk 5]
```diff
@@ -1311,7 +1311,53 @@ func (s *Session) ChannelMessageSendComplex(channelID string, data *MessageSend)
...
+		if err != nil {
+			return
+		}
+
+		if err = bodywriter.WriteField("tts", strconv.FormatBool(data.Tts)); err != nil {
```
Comment (iopred) [inline restapi.go]: Probably only if tts is true.

(also, line 1337 👍  )

File: restapi.go (hunk 3 above)
Comment (legolord208) [inline restapi.go]: I felt some sending, even if useless, parameters wasn't the biggest deal. I'll fix it though!

File: restapi.go [hunk 6]
```diff
@@ -1311,7 +1311,59 @@ func (s *Session) ChannelMessageSendComplex(channelID string, data *MessageSend)
...
+
+		if data.Embed != nil {
+			var embed []byte
+			embed, err = json.Marshal(data.Embed)
+			if err != nil {
+				return
+			}
+			err = bodywriter.WriteField("embed", string(embed))
+			if err != nil {
+				return
```
Comment (legolord208) [inline restapi.go]: Pointing out line 1337, @iopred? `return` where you came from... :rofl:

**Pull request merged**